xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -249,30 +256,75 @@ public void updateLocalityStore(Map<XdsLocality, LocalityInfo> localityInfoMap)
         }
         newState = aggregateState(newState, childHelper.currentChildState);
       }
-      localityMap = Collections.unmodifiableMap(updatedLocalityMap);
 
-      updatePicker(newState, childPickers);
+      // Add deactivated localities to localityMap to keep track of them
+      for (XdsLocality locality : oldLocalities) {
+        if (localityMap.get(locality).isDeactivated()) {
+          updatedLocalityMap.put(locality, localityMap.get(locality));
+        }
+      }
+      localityMap = updatedLocalityMap.build();
 
+      final Set<XdsLocality> toBeRemovedFromStatsStore = new HashSet<>();
       // There is a race between picking a subchannel and updating localities, which leads to
       // the possibility that RPCs will be sent to a removed locality. As a result, those RPC
       // loads will not be recorded. We consider this to be natural. By removing locality counters
       // after updating subchannel pickers, we eliminate the race and conservatively record loads
       // happening in that period.
+      for (XdsLocality oldLocality : edsResponsLocalityInfo.keySet()) {
+        if (!localityInfoMap.containsKey(oldLocality)) {
+          toBeRemovedFromStatsStore.add(oldLocality);
+        }
+      }
       helper.getSynchronizationContext().execute(new Runnable() {
         @Override
         public void run() {
-          for (XdsLocality locality : toRemove) {
+          for (XdsLocality locality : toBeRemovedFromStatsStore) {
             loadStatsStore.removeLocality(locality);
           }
         }
       });
+
+      edsResponsLocalityInfo = ImmutableMap.copyOf(localityInfoMap);
+      updatePicker(newState, childPickers);
     }
 
     @Override
     public void updateDropPercentage(ImmutableList<DropOverload> dropOverloads) {
       this.dropOverloads = checkNotNull(dropOverloads, "dropOverloads");
     }
 
+    private void deactivate(final XdsLocality locality) {
+      if (!localityMap.containsKey(locality) || localityMap.get(locality).isDeactivated()) {
+        return;
+      }
+
+      final LocalityLbInfo localityLbInfo = localityMap.get(locality);
+      class DeletionTask implements Runnable {
+
+        @Override
+        public void run() {
+          localityLbInfo.shutdown();
+
+          ImmutableMap.Builder<XdsLocality, LocalityLbInfo> builder = ImmutableMap.builder();
+          for (Map.Entry<XdsLocality, LocalityLbInfo> entry : localityMap.entrySet()) {
```

Comment:
Isn't this too expensive? Maybe should not make `localityMap` immutable so that we can do add/remove easily. It is not exposed to external classes, so I think it is fine to use mutable map internally.

Reply:
We started to create new map each update to make it order-deterministic https://github.com/grpc/grpc-java/commit/44bbccff79d108baf90f55226a4fbc62fcbcaab4#diff-6c00f5da836b0a8dcca3eee70168c19b

This adds overhead for GC but assuming the map is not large and the update wont be so frequent (not like 1/millis), GC can easily handle it.

Reply:
Sure. If performance it not our concern now. Although it also makes the implementation clearer.